### PR TITLE
matrix-synapse: 1.9.1 -> 1.11.1

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

This updates `matrix-synapse` to [1.11.1](https://github.com/matrix-org/synapse/releases/tag/v1.11.1).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Ma27** [comment]
@GrahamcOfBorg test matrix-synapse postgresql

**Comment by jonringer** [comment]
I fixed pythonPackages.signedjson https://github.com/NixOS/nixpkgs/pull/80449, not sure if you just want to pull the changes into this pr or not

**Review by fadenb — APPROVED** [review]

**Review by mguentner — APPROVED** [review]
Thanks!
Anyone got an idea how this can be backported to 19.09 or any following release for that matter as we would add some changes to the postgresql module which might not be desired.

**Comment by Ma27** [comment]
>  Anyone got an idea how this can be backported to 19.09 or any following release for that matter as we would add some changes to the postgresql module which might not be desired.

I'm actually using this on 19.09 already (my matrix HS is still running 19.09 on a custom branch based on `release-19.09` with this change cherry-picked on top) and the database still works fine (you get a warning at the beginning though), so it should be fine (considering the fact that 19.09 will only live for a few months at most).

**Review by Ekleog — APPROVED** [review]
I think the short lifetime of 19.09 is actually an argument in favor of not backporting there, as the risk of breakage with a revert looks higher than the risk of breakage without a revert IMO

**Comment by Ekleog** [comment]
(however, I do think it totally makes sense to backport to 20.03, because even though we're technically past cut-off time, we haven't said it's stabilized yet, and the risk of breakage in the following 9 months is quite high, given we're speaking of synapse)

**Comment by Ma27** [comment]
> (however, I do think it totally makes sense to backport to 20.03, because even though we're technically past cut-off time, we haven't said it's stabilized yet, and the risk of breakage in the following 9 months is quite high, given we're speaking of synapse)

Just for the record: we backported every release of matrix-synapse to stable in the past (at least since I'm taking care of the package) as many releases either contained important bugfixes or were needed due to some infrastructure changes at matrix.org that would break a self-hosted synapse otherwise.

**Review by fpletz — APPROVED** [review]

File: nixos/modules/services/databases/postgresql.nix
Comment (grahamc) [inline nixos/modules/services/databases/postgresql.nix]: I'm not a fan of these options. I think `ensure` is in general a "bad word" for a declarative tool like NixOS. I think Nix isn't a good tool for this type of mechanism. @aanderse and @flokli and I have talked in the past about supporting this in a way similar to the "users" support (mutableUsers / creation / deletion as needed.)

In particular, I'm uncomfortable with these `lc_collate` and `lc_ctype` options because they are not declarative, and don't _ensure_ these databases _are_ as specified. In other words, this is creating a type of untracked state where the configuration.nix is lying about the system, something I think we should try hard to avoid.

Concretely, these two systems will end up in different states, something NixOS typically tries very hard to do:

```
+ ensureDatabases.foo.lc_collate = "foo";
$ nixos-rebuild switch
- ensureDatabases.foo.lc_collate = "foo";
+ ensureDatabases.foo.lc_collate = "bar";
$ nixos-rebuild switch
```


```
+ ensureDatabases.foo.lc_collate = "bar";
$ nixos-rebuild switch
```

I would prefer to not see this part of this PR merge.

Comment (Ma27) [reply nixos/modules/services/databases/postgresql.nix]: First of all, thanks a lot for raising your concerns about those patches!

> I think ensure is in general a "bad word" for a declarative tool like NixOS

Those options were introduced earlier (we have `ensure{Users,Databases}` for quite a while now). Also, I'd like to mention that previously options such as [`initialScript`](https://nixos.org/nixos/options.html#services.postgresql.initialscript) were used that were even more impure.

Or in other words: do you have a suggestion how to manage that kind of state in Nix? (Admittedly I'm fairly unsafe when it comes to state in Nix, I mostly use workarounds like those `ensure*`-options or `initialScript` as I'm currently not aware of a better solution).

> @aanderse and @flokli and I have talked in the past about supporting this in a way similar to the "users" support (mutableUsers / creation / deletion as needed.)

In case this was discussed in IRC - do you have a link to that discussion? :) 

> In particular, I'm uncomfortable with these lc_collate and lc_ctype options because they are not declarative, and don't ensure these databases are as specified. In other words, this is creating a type of untracked state where the configuration.nix is lying about the system, something I think we should try hard to avoid.

As already mentioned, I'm rather uncomfortable when it comes to actual state-management with Nix. May I ask if you have a different suggestion (don't get me wrong - I don't expect you to know "the perfect" solution just because you raised concerns about those patches. I'm just curious as I currently don't know any better fix than the one I filed here).

Comment (grahamc) [reply nixos/modules/services/databases/postgresql.nix]: I know `ensurexxx` has been here longer than this PR, and I don't like it any better :P. ensureUsers in particular is Very scary for me. If initialScript could work, I think that would be okay-ish. initialScript is less bad, since it has in its name the fact that it is almost guaranteed to be lies and doesn't attempt to pretend that it is declarative.

Comment (Ma27) [reply nixos/modules/services/databases/postgresql.nix]: So your main problem are those `ensure*` options, especially if there are additional options to configure a database or a user as those can be wrongly interpreted as declarative? And that the same issue exists with the GRANT-options for `ensureUsers`?

> If initialScript could work, I think that would be okay-ish

I'm not sure if I can agree here. I still remember the first time I used the postgresql-module a few years ago and it took me a while until I figured out that `initialScript` is only executed if `${cfg.dataDir}/.first_startup` doesn't exist. What I want to say is that this can be misinterpreted as well, but in the end the behavior of both `initialScript` and `ensure*` is currently well-documented and doesn't affect existing setups (at least in the case of matrix-synapse).

If you insist on not enhancing and using `ensure*` here, we can still revert #75294 and fix the issue in the systemd-unit.

Otherwise I personally think that this is a general issue with the way Nix "handles" state and further measures should be discussed in a separate issue.

Comment (pacien) [reply nixos/modules/services/databases/postgresql.nix]: `ensureDatabases.*.initializationOptions.*` would be both more convenient and explicit enough in my opinion.

Comment (mguentner) [reply nixos/modules/services/databases/postgresql.nix]: The PR is more about `synapse` than `ensureDatabases`. However the later blocks the PR. How can we resolve this deadlock?

Comment (Ma27) [reply nixos/modules/services/databases/postgresql.nix]: @mguentner I basically agree with you (that's why I suggested to merge this and discuss the issues about `ensure*` in a separate issue), however I certainly understand @grahamc's concerns about worsening the situtation in `nixos/postgresql` and I'd like to have this resolved before having those patches in upstream.

We'll probably talk about the entire issue synchronously on IRC and try to find a suitable comproise.

Apart from this I'd probably add `matrix-synapse` at [`1.11`](https://github.com/matrix-org/synapse/releases/tag/v1.11.0) before considering this mergable again.

**Comment by Ma27** [comment]
Just updated matrix-synapse to the latest upstream version (1.11.0).

File: nixos/modules/services/databases/postgresql.nix
Comment (florianjacob) [reply nixos/modules/services/databases/postgresql.nix]: Very important point about `lc_ctype` and `lc_collate` violating the current behaviour of `ensure*`, and being more similar to `initialScript`, though with database creation level instead of database software installation level granularity.
As synapse (may?) will eventually disallow databases with incorrect ctype and collation, don't we need some form of automatic ctype & collation migration anyway? That would allow to have a `lc_ctype` and `lc_collate` option that doesn't lie about the system state and behaves with the same guarantees like all other `ensure*` options.

Comment (Ma27) [reply nixos/modules/services/databases/postgresql.nix]: I reverted the patch that used ensure* in the matrix-synapse module, so I consider this resolved for now.

> don't we need some form of automatic ctype & collation migration anyway

IMHO no: as we've seen e.g. throughout this discussion Nix doesn't do a very good job at state management. IMHO the safest approach here would be to ask administrators to manually do the migration.

Please note that I just realized that this only affects synapse setups from the last few months as we configured databases properly before switching to ensure* here (b2d67c08d03679ba1b9a928ac5610ad940846e7f).

File: nixos/doc/manual/release-notes/rl-2003.xml
Comment (Ma27) [inline nixos/doc/manual/release-notes/rl-2003.xml]: Note to self: as soon as we have enough approvals here I should update the release notes.

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [inline nixos/modules/services/misc/matrix-synapse.nix]: This will cause issues with setups that were created on nixos-unstable while ensureDatabases were used as in those cases a database was created with touching `${cfg.dataDir}/db-created`.

Any ideas atm how we can work around this?

**Review by kampka — COMMENTED** [review]
It would be nice for further maintenance if the commit message would contain information about why the revert was necessary. It's generally hard to pinpoint such things on the issue tracker from just the commit history I find.

**Review by kampka — APPROVED** [review]

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (mguentner) [reply nixos/modules/services/misc/matrix-synapse.nix]: @kampka Suggested to use an idempotent SQL statement. This could determine whether the database exists and create it along with the user and role. After the execution, `db-created` will be created, either because the script determined that nothing is to do and exited or because the script created the resources.

sth like
```
SELECT COUNT(*) FROM pg_database WHERE datname="matrix-synapse";
```

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: Wouldn't we reintroduce the issues @grahamc expressed with `ensure*`?

Comment (aanderse) [reply nixos/modules/services/misc/matrix-synapse.nix]: @Ma27 I don't understand what you mean by "reintroduce" - the issues @grahamc brought up have not been resolved by any approach mentioned so far.

Comment (mguentner) [reply nixos/modules/services/misc/matrix-synapse.nix]: This PR is now open since 19 days and stalled because of @grahamc 's opinion on `ensureDatabases`. 

> I would prefer to not see this part of this PR merge.
-- @grahamc 

This is clearly resolved. We cannot guarantee the state of postgresql. State may change outside of Nix's control, whether we like it or not. We will and should not solve this issue here.
Can we please get this into master and backport this to 19.09 already?

@Ma27 I will ping you concerning the SQL stanza.

**Comment by Ma27** [comment]
Let me try to summarize the issue here and propose some solutions:

`matrix-synapse` 1.10 requires `LC_CTYPE` and `LC_COLLATE` to be set to `C` when using a postgresql database for persistence (which basically means that postgresql should ignore any locale-preferences: https://www.postgresql.org/docs/12/locale.html). This only affects new databases, existing databases will remain usable, but `matrix-synapse` will start with a warning.

I'm also pretty sure that the majority of setups isn't affected by this since the database was configured properly in 19.09 and older. Since b2d67c08d03679ba1b9a928ac5610ad940846e7f the `matrix-synapse` module uses the `ensure*` options that don't configure the database properly (this change is only on `nixos-unstable`).

My original solution added the `lc_{ctype,collate}` options to the `ensureDatabases` option in the `postgresql`-module. As @grahamc noted this has the downside that we break declarativity here: the options won't be taken into account if the database already exists, so if we want to "ensure" a database to exist, but it already exists, the database won't be configured properly. I'd also like to mention that I'm pretty much against automated migration since Nix isn't designed to handle application's state at the moment and there's a lot of stuff that can go wrong (e.g. when using `nixops deploy --allow-reboot`).

I can currently think of the following solutions:

* drop database management entirely and do the setup in the test. This will only affect new setups as existing setups already have an existing database. This is already done for `nextcloud`, `codimd`, `slurm` and a few more (my preferred solution atm).

* Use `initialScript` in the module (this is probably a bad idea since the database won't be configured if `matrix-synapse` is deployed to a machine with an already existing postgresql since `initialScript` will only be used at the *very* first startup of postgresql).

* Leave it as it is. This has the downside that new matrix-synapse setups will probably have trouble upgrading since the database-creation will be done again since `/var/lib/matrix-synapse/db-created` doesn't exist.

I think the best solution is to remove the database configuration entirely for now.

@mguentner @fpletz @fadenb @aanderse @grahamc @kampka  thoughts on this?

**Comment by ptman** [comment]
Maybe the declarativity can be helped with naming? Like using `create_time_lc_collate` instead of `lc_collate` or something similar that signifies that it only affects new databases.

**Comment by pacien** [comment]
> * Leave it as it is. This has the downside that new matrix-synapse setups will probably have trouble upgrading since the database-creation will be done again since `/var/lib/matrix-synapse/db-created` doesn't exist.

Instead of checking the existence of that flag file, we could check the existence of the role and database directly and initialise those only if they don't, in some idempotent transaction executed in the preStart phase as @aanderse suggested.  This should be compatible with all the previous automatic initialisation methods used in this module in the past.

**Review by fadenb — APPROVED** [review]

**Comment by Ma27** [comment]
I just discussed the issue with @fpletz today. We decided that the best solution for now is to remove the database handling entirely (as it's implemented in several other modules). The change will be documented and users can copy configurations from the VM test or the docs for a database setup.

New users will have to take care of their database on their own, users with existing setups won't have any issues as the database is configured already.

If you deploy an existing configuration to a new machine, the missing database-setup will be detected by an assertion which instructs the user to either configure the database manually or to provide a database on a foreign host.

The change is currently work in progress :)

**Comment by Ma27** [comment]
Just had a short chat with @aanderse this afternoon about this where we both agreed that it's the best solution for now to drop the database configuration since Nix lacks proper safety measures for application-state handling.

Additionally, if we add an idempotent setup later, it shouldn't be too complicated to add this to the `matrix-synapse` module IMHO.

@GrahamcOfBorg test matrix-synapse

@grahamc after several discussions I decided that it's best for now to drop the database configuration for now. I should also mention that I fully agree with your concerns after thinking a little bit more about this :) 

@fpletz and I think that we should merge (and backport) this soon though since this fixes an actual security issue of `matrix-synapse`.

File: nixos/doc/manual/release-notes/rl-2003.xml
Comment (Ma27) [reply nixos/doc/manual/release-notes/rl-2003.xml]: Release-notes were updated during the latest push.

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: We'll drop the database setup entirely for now. An example is shown in the vm-test, users with existing setups won't have any problems since they already have a fully configured database. For further information, please read the release-notes I added in this PR.

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: As mentioned in https://github.com/NixOS/nixpkgs/pull/80447#discussion_r392385305, the db setup will be dropped, so this doesn't need to be discussed anymore.

File: nixos/doc/manual/release-notes/rl-2003.xml
Comment (aanderse) [inline nixos/doc/manual/release-notes/rl-2003.xml]: This SQL will only execute if `postgresql` has never ran before. It seems possible this documentation wouldn't work for a at least a number of users.

Comment (Ma27) [reply nixos/doc/manual/release-notes/rl-2003.xml]: This is actually the reason why I linkified the `initialScript` option here :)

File: nixos/doc/manual/release-notes/rl-2003.xml
Comment (fpletz) [inline nixos/doc/manual/release-notes/rl-2003.xml]:
```suggestion
     Due to <link xlink:href="https://github.com/matrix-org/synapse/releases/tag/v1.10.0rc1">stricter requirements</link>
```

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (fpletz) [inline nixos/modules/services/misc/matrix-synapse.nix]: Indentation gone wrong here?

**Review by fpletz — APPROVED** [review]

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: I did this on purpose: the first line will be prefixed with a `- `, so with my indention the error message will actually look like this:

```
error:
Failed assertions:
- Cannot deploy matrix-synapse with a configuration for a local postgresql database
  and a missing postgresql service. Since 20.03 it's mandatory to manually configure the
  database (please read the thread in https://github.com/NixOS/nixpkgs/pull/80447 for
  further reference).

  If you
  - try to deploy a fresh synapse, you need to configure the database yourself. An example
    for this can be found in <nixpkgs/nixos/tests/matrix-synapse.nix>
  - update your existing matrix-synapse instance, you simply need to add `services.postgresql.enable = true`
    to your configuration.

For further information about this update, please read the release-notes of 20.03 carefully.

(use '--show-trace' to show detailed location information)
```

**Comment by Ma27** [comment]
@GrahamcOfBorg test matrix-synapse

**Review by mguentner — APPROVED** [review]
LGTM

**Review by kampka — APPROVED** [review]

**Comment by Ma27** [comment]
Please don't merge it yet. @florianjacob made some good suggestions about how we should document this and I'd like to add this (already started working on it) first :)

**Comment by Ma27** [comment]
@GrahamcOfBorg test matrix-synapse

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (mguentner) [inline nixos/modules/services/misc/matrix-synapse.nix]: The only remark: If a config has `postgresql.enable = true` for another service, this message will not be shown.
Only a check of the system logs will point to the missing database. We cannot check for everything :woman_shrugging:

**Review by mguentner — APPROVED** [review]
Nice documentation. Looks good to me.

File: nixos/modules/services/misc/matrix-synapse.nix
Comment (Ma27) [reply nixos/modules/services/misc/matrix-synapse.nix]: Well, one of the reasons for this assertion is to ensure that people with an existing matrix-synapse and a local database don't deploy a config without a postgresql service (that isn't enabled anymore by this module).

**Comment by Ma27** [comment]
Just pushed a change which fixes a minor typo in the docs. As soon as the eval is finished, I'll merge this now. The latest changes from Sunday were documentation-related (and reviewed by @mguentner), the actual update has been discussed pretty much now and should be mergable.

Also, I'm running matrix-synapse 1.11.1 on NixOS 19.09 for a while now without any issues, the change should be fairly safe to use (and to backport).

Thanks a lot for everyone involved here!

**Comment by aanderse** [comment]
Thanks for all your work on this @Ma27 :tada:

**Comment by Ma27** [comment]
Backported to stable:

* 20.03: 7f8e302fb6513d1d5b59f22a9948a5610ca75296, ba6271a49c6079e0223ee9a166d6a65b0de05445, 23139ff62fb08f7bf7b3a4efc30e6e798d9f7e26
* 19.09: a9d474652f740fbdbf4c0e29158e26508d17a01a, dce33f1a2d491489ad044259c7d06d8870a205fb

**Comment by mguentner** [comment]
Thanks for everyone involved, especially @Ma27

**Comment by luochen1990** [comment]
Could you please have a look at this PR https://github.com/NixOS/nixpkgs/pull/339195 ?  I need some help since I haven't write a unit test for postgresql (especially in nixpkgs) before

**Pull request merged**
